minecraft_go/software/mc_go/src/gps.c

```diff
@@ -129,9 +145,249 @@ int gps_get_gga_data(char *GGA_line, struct gga_data *holder){
+		token = strtok(NULL, delimiter);
+
+		count++;
+	}
+
+	return true;
+}
+
+
+/*
+ *  Get the most up to date RMC data. Return true if succeed before max_tries.
+ *
+ *  max_tries param is intend to prevent program from staling when GPS is not
+ *  working and a value of 20 should be enough unless we are in the middle of
+ *  a large data dump.
+ */
+bool getCurrentRMCdata(RMC_data *buffer, int max_tries){
+	char *data_line_buffer = malloc(GPS_DEFAULT_DATA_LINE_SIZE * sizeof(char));
+
+	int num_tries = 0;
+	do {
+		num_tries++;
+		gps_retrieve_data_line(data_line_buffer, GPS_DEFAULT_DATA_LINE_SIZE);
+	} while(!gps_get_rmc_data(data_line_buffer, buffer) && num_tries < max_tries);
+
+	free(data_line_buffer);
+
+	return num_tries >= max_tries ? false : true;
+}
+
+//---------------------------Timer Functions------------------------------//
+
+void convertRMCtoDateTime(RMC_data *RMC_data, DateTime *buffer){
+	char RMC_year[3] = {0};
+	strncpy(RMC_year, RMC_data->date+4, 2);
+	buffer->year = atoi(RMC_year);
+
+	char RMC_month[3] = {0};
+	strncpy(RMC_month, RMC_data->date+2, 2);
+	buffer->month = atoi(RMC_month);
+
+	char RMC_day[3] = {0};
+	strncpy(RMC_day, RMC_data->date, 2);
+	buffer->day = atoi(RMC_day);
+
+	char RMC_hour[3] = {0};
+	strncpy(RMC_hour, RMC_data->UTC_time, 2);
+	buffer->hour = atoi(RMC_hour);
+
+	char RMC_minute[3] = {0};
+	strncpy(RMC_minute, RMC_data->UTC_time+2, 2);
+	buffer->minute = atoi(RMC_minute);
+
+	char RMC_second[3] = {0};
+	strncpy(RMC_second, RMC_data->UTC_time+4, 2);
+	buffer->second = atoi(RMC_second);
+}
+
+
+/*
+ *  Return elapsed time in seconds from start till finish
+ *
+ *  For simplicity, assume each year is 365 days, month is 30 days for now.
+ */
+unsigned long getElapsedInSeconds(DateTime *start, DateTime *finish){
+	unsigned long start_in_seconds;
+	unsigned long finish_in_seconds;
+
+	start_in_seconds = start->year * 31536000 +
+					   start->month * 2592000 +
+					   start->day * 86400 +
+					   start->hour * 3600 +
+					   start->minute * 60 +
+					   start->second;
+
+	finish_in_seconds = finish->year * 31536000 +
+					    finish->month * 2592000 +
+					    finish->day * 86400 +
+					    finish->hour * 3600 +
+					    finish->minute * 60 +
+					    finish->second;
+
+	return finish_in_seconds - start_in_seconds;
+}
+
+void convertSecondsToTime(Time *buffer, unsigned long seconds){
+	unsigned long left_over;
+
+	buffer->hour = seconds / 3600;
+	left_over = seconds % 3600;
+	buffer->minute = (int) left_over / 60;
+	buffer->second = (int) left_over % 60;
+}
+
+void startTimer(DateTime *start_time){
+	RMC_data *RMC_buffer = malloc(sizeof(RMC_data));
+
+	if (!getCurrentRMCdata(RMC_buffer, GPS_DEFAULT_DATA_RETRIEVAL_TRIES))
+		printf("Error: Unable to retrieve RMC data in %d tries.\n", GPS_DEFAULT_DATA_RETRIEVAL_TRIES);
+	else
+		convertRMCtoDateTime(RMC_buffer, start_time);
+
+	free(RMC_buffer);
+}
+
+unsigned long stopTimer(DateTime *start_time){
+	RMC_data *RMC_buffer = malloc(sizeof(RMC_data));
```

Comment:
is it possible to declare these on the stack instead of using malloc? Less overhead
```c
RMC_data RMC_buffer;
//
convertRMCtoDateTime(&RMC_buffer);
```